### PR TITLE
Prow script for pilot-multicluster-e2e test

### PR DESCRIPTION
**Pull request by tiswanso**
*NOTE:* New PR for closed PR #5516 -- see that PR for prior review comments

Adds the script called by CI to run the pilot multicluster e2e test. Uses mason to get GKE resources in the same GCP project. The boskos resource_type already creates multiple GKE clusters. Uses kubectl config get-contexts to determine all k8s clusters to use for the multicluster env.

Summary:
*  Invokes the `make e2e_pilotv2_v1alpha3` test via `e2e-suite.sh` with the pilot e2e multicluster config flag `cluster_registry_dir` 
*  The script can be run with pre-allocated clusters by having a pre-setup
kubeconfig for multiple clusters. It will choose the pilot cluster to
be the first cluster in the kubectl config get-contexts or the context set in the PILOT_CLUSTER env var.
*  The script creates a service-account and corresponding kubeconfig for the remote cluster and passes to the pilot e2e test via the `cluster_registry_dir`
*  For GKE, firewall rules are created to allow intercluster pod-to-pod reachability
*  Timeout for multicluster pilot test raised to 35 minutes (seen in practice).
*  Cleanup actions:
   *  test created GKE firewall rules deleted
   *  remote cluster service-account deleted

Example for use with 2 precreated GKE clusters and designating "gke_istio-ci-testing_us-east1-c_cluster-1" context as the cluster to host the istio control-plane (pilot, etc).

```
PILOT_CLUSTER=gke_istio-ci-testing_us-east1-c_cluster-1 \
  KUBE_USER=tiswanso@cisco.com ROOT=$(pwd) USE_MASON_RESOURCE=False \
  prow/istio-pilot-multicluster-e2e.sh
```

for non-GKE clusters:
```
PILOT_CLUSTER=cluster-1 USE_GKE=False \
  KUBE_USER=tiswanso@cisco.com ROOT=$(pwd) USE_MASON_RESOURCE=False \
  prow/istio-pilot-multicluster-e2e.sh
```

### CODE REVIEW

File: prow/cluster_lib.sh [hunk 1]
```diff
@@ -20,23 +20,147 @@
...
+    SERVICE_ACCOUNT=istio-multi-test
+
+    # mason dumps all the kubeconfigs into the same file but we need to use per cluster
+    # files for the clusterregsitry config.  Create the separate files.
+    #  -- if PILOT_CLUSTER not set, assume pilot install to be in the first cluster
+    PILOT_CLUSTER=${PILOT_CLUSTER:-$(kubectl config current-context)}
+    unset IFS
+    k_contexts=$(kubectl config get-contexts -o name)
+    for context in ${k_contexts}; do
+        if [[ "$PILOT_CLUSTER" != "$context" ]]; then
```
Comment (sebastienvas) [inline prow/cluster_lib.sh]: {}

File: prow/cluster_lib.sh [hunk 2]
```diff
@@ -20,23 +20,147 @@
...
+         - name: ${CLUSTER_NAME}
+           user:
+             token: ${TOKEN}
+EOF
+
+}
+
+function setup_clusterreg () {
+    # setup cluster-registries dir setup by mason
+    CLUSTERREG_DIR=${CLUSTERREG_DIR:-$(mktemp -d /tmp/clusterregXXX)}
```
Comment (sebastienvas) [inline prow/cluster_lib.sh]: Please add "" around variable.
CLUSTERREG_DIR="${CLUSTERREG_DIR:-$(mktemp -d /tmp/clusterregXXX)}"

File: prow/cluster_lib.sh [hunk 3]
```diff
@@ -20,23 +20,147 @@
...
+  if [[ "${USE_GKE}" == "True" && "${SETUP_CLUSTERREG}" == "True" ]]; then
+    ALL_CLUSTER_CIDRS=
+    for cidr in $(gcloud container clusters list --format='value(clusterIpv4Cidr)'); do
+      if [[ "${ALL_CLUSTER_CIDRS}" != "" ]]; then
+        ALL_CLUSTER_CIDRS+=','
+      fi
+      ALL_CLUSTER_CIDRS+=${cidr}
+    done
+    ALL_CLUSTER_NETTAGS=
+    for net_tag in $(gcloud compute instances list --format='value(tags.items.[0])'); do
```
Comment (sebastienvas) [inline prow/cluster_lib.sh]: gcloud compute instances list --format='value(tags.items.[0] | sort | uniq ?

Comment (sebastienvas) [reply prow/cluster_lib.sh]:
```bash
function join_by { local IFS="$1"; shift; echo "$*"; }

ALL_CLUSTER_CIDRS=($(gcloud container clusters list --format='value(clusterIpv4Cidr)' | sort | uniq))
ALL_CLUSTER_CIDRS="${join_by , ${TEST}"
```

**Comment by tiswanso** [comment]
@sebastienvas -- updating with your comments, thx!

File: prow/cluster_lib.sh (hunk 3 above)
Comment (tiswanso) [reply prow/cluster_lib.sh]: Yes... `sort | uniq` is much nicer!  fixing for this and ALL_CLUSTER_NETTAGS

File: prow/istio-pilot-multicluster-e2e.sh [hunk 4]
```diff
@@ -16,10 +16,10 @@
 
 # set this to the new multicluster-e2e type
 export RESOURCE_TYPE="${RESOURCE_TYPE:-gke-e2e-test}"
-export OWNER=istio-pilot-multicluster-e2e
+export OWNER="istio-pilot-multicluster-e2e"
 
 export SETUP_CLUSTERREG="True"
-CLUSTERREG_DIR=${CLUSTERREG_DIR:-$(mktemp -d /tmp/clusterregXXX)}
+CLUSTERREG_DIR="${CLUSTERREG_DIR:-$(mktemp -d /tmp/clusterregXXX)}"
```
Comment (sdake) [inline prow/istio-pilot-multicluster-e2e.sh]: might consider more xxx - my understanding is prow is a shared machine and you may run into collisions with such a small address space.

Comment (tiswanso) [reply prow/istio-pilot-multicluster-e2e.sh]: this is running in a container spawned every run by prow.  IMO, it's safe as is.

**Comment by tiswanso** [comment]
Hi @sdake @sebastienvas @costinm -- can we hold the nit comments for a follow-on so we can get this in ASAP? <waving white flag>   We need to get this in for multicluster testing via prow.  

history:
This script has been in review for months at this point (see #5516), it was approved but not allowed into rel 1.0, went stale, closed by bot, and had to be reopened under this one.  I've got about 4 diff sets on top of this for multicluster testing and every change to this prow set takes me about 45min+ to retest and resubmit because of all the other dependent unmerged change in the e2e tests.  This functions.  I test with it almost every day with various multicluster scenarios for the rel1 testing.

File: prow/istio-pilot-multicluster-e2e.sh (hunk 4 above)
Comment (sebastienvas) [reply prow/istio-pilot-multicluster-e2e.sh]: @tiswanso  is right. Every job run in a container.

**Comment by tiswanso** [comment]
rebased after rel1.0 merge

**Comment by tiswanso** [comment]
last change was just a rebase to get codecov passing after merge of codecov new baseline

**Comment by john-a-joyce** [comment]
/lgtm 
can we try to get this merged ASAP so that we can finally enable the e2e tests

**Pull request merged**